### PR TITLE
feat: add tracing support

### PR DESCRIPTION
**Pull request by musale**
Part of https://github.com/microsoft/kiota/issues/1872

- [x] `send_async`
- [x] `send_collection_async`
- [x] `send_collection_of_primitive_async`
- [x] `send_primitive_async`
- [x] `send_no_response_content_async`
- [x] `convert_to_native_request_async`
- [x] middleware module

### CODE REVIEW

File: kiota_http/middleware/url_replace_handler.py [hunk 1]
```diff
@@ -30,12 +31,17 @@ async def send(
         Returns:
             Response: The response object.
         """
+        response = None
+        _enable_span = self._create_observability_span(request, "UrlReplaceHandler_send")
+        _enable_span.set_attribute("com.microsoft.kiota.handler.url_replacer.enable", True)
         current_options = self._get_current_options(request)
 
         url_string: str = str(request.url)  #type: ignore
         url_string = self.replace_url_segment(url_string, current_options)
```
Comment (baywet) [inline kiota_http/middleware/url_replace_handler.py:40]: the replacement and the span attribute being set to true should only happen if the option is enabled.

File: kiota_http/_exceptions.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+"""Exceptions raised in Kiota HTTP."""
+
+
+class KiotaHTTPXError(Exception):
+    """Base class for Kiota HTTP exceptions."""
+
+
+class BackingstoreError(KiotaHTTPXError):
```
Comment (baywet) [inline kiota_http/_exceptions.py:8]: I'm not sure why we're introducing specialized error types here?

**Review by baywet — CHANGES_REQUESTED** [review]

File: kiota_http/_exceptions.py (hunk 2 above)
Comment (musale) [reply kiota_http/_exceptions.py:8]: I added them to get specialized context of the source of the error as well as the kind of the error. Previously we just raised and `Exception(msg)` which is the basic way to do it. However, it limits handling the errors. Specializing the errors for the lib offers:
- ability for the client to catch specific errors. i.e., in tests or when using the library.
- A clear source and kind of error in the stack trace. A `RedirectError` is most probably raised from the `redirect_handler` module and its value will most probably state the kind of redirect error.
- ability to catch _all_ errors under the base error class `KiotaHTTPXError`. So, all the subclassed errors can easily be handled in a single block (if needed by the lib user).

IMO, it's a good practice which goes a long way in developer usage of the lib especially it being a core library.

Comment (baywet) [reply kiota_http/_exceptions.py:8]: I just saw this comment now, sorry about the late reply. I agree that specialized error types make handling easier in most languages. Couple of challenges here:
- This error is related to the backing store, which is a concept defined in the abstractions, wouldn't it make more sense to have the error defined in abstractions?
- The name has a casing issue
- Python would be the only language doing that to date

Comment (musale) [reply kiota_http/_exceptions.py:8]: Oh, I see your point @baywet. For point 1 and 3, the naming is to aid with the raised exception when using the library in Python. A developer will get a custom error name. I believe the experience is different in other languages. Maybe, we can change the name to something more inclined to `Kiota http` like `KiotaHttpBackingStoreError`? For point 2 I don't get the challenge?

Comment (baywet) [reply kiota_http/_exceptions.py:8]: BackingstoreError => BackingStoreError (note the capital S)

**Review by baywet — DISMISSED** [review]

**Comment by samwelkanda** [comment]
Failing test depends on https://github.com/microsoft/kiota-abstractions-python/pull/136

**Review by samwelkanda — DISMISSED** [review]

**Review by samwelkanda — APPROVED** [review]

**Pull request merged**